service/history/historyEngine.go

```diff
@@ -1502,9 +1506,11 @@ Update_History_Loop:
 				}
 
 				if ai.StartedID == common.EmptyEventID {
-					// We haven't started the activity yet, we can cancel the activity right away.
+					// We haven't started the activity yet, we can cancel the activity right away and
+					// schedule a decision task to ensure the workflow makes progress.
 					msBuilder.AddActivityTaskCanceledEvent(ai.ScheduleID, ai.StartedID, *actCancelReqEvent.EventId,
 						[]byte(activityCancelationMsgActivityNotStarted), common.StringDefault(request.Identity))
+					activityNotStartedCancelled = true
```

Comment:
have to use a separate flag here, reusing `hasUnhandledEvents` causes issues if we send also have a complete in our set of decisions.